addons/pos_self_order/static/src/router.js

```diff
@@ -0,0 +1,94 @@
+                    return [name, parseInt(match)];
+                case "string":
+                    return [name, match];
+                default:
+                    throw new Error(`Unknown type ${type}`);
+            }
+        })
+    );
+}
+export class Router extends Component {
+    static template = xml`<t t-slot="{{state.activeSlot}}" t-props="state.slotProps"/>`;
+    setup() {
+        useSubEnv({ navigate: this.navigate.bind(this) });
+        this.state = useState({
+            activeSlot: "default",
+            slotProps: {},
+        });
+        window.addEventListener("self_order_navigate", () => {
+            this.matchURL();
+        });
+        // this is needed for the back button to work
+        window.addEventListener("popstate", (event) => {
+            event.preventDefault();
+            this.matchURL();
+        });
+        this.routes = Object.keys(this.props.slots).map((route) => {
+            const paramStrings = route.match(/\{\w+:\w+\}/g);
+            if (!paramStrings) {
+                return { route, paramSpecs: [], regex: new RegExp(`^${route}$`) };
+            }
+            const paramSpecs = paramStrings.map((paramString) => {
+                const [, type, name] = paramString.match(/(\w+):(\w+)/);
+                return { type, name };
+            });
+            const regex = new RegExp(
+                `^${route
+                    .split(/\{\w+:\w+\}/)
+                    .map((part) => escapeRegExp(part))
+                    .join("(.*)?")}$`
+            );
+            return { route, paramSpecs, regex };
+        });
+        this.matchURL();
+    }
+
+    matchURL() {
+        const path = browser.location.pathname;
+        for (const { route, paramSpecs, regex } of this.routes) {
+            const match = path.match(regex);
+            if (match) {
+                const parsedParams = parseParams(match.slice(1), paramSpecs);
+                this.state.activeSlot = route;
+                this.state.slotProps = parsedParams;
+                return;
+            }
+        }
+        this.state.activeSlot = "default";
+        this.state.slotProps = {};
+    }
+    /**
+     * @param {string} route
+     * @description Navigate to the given route.
+     * We can have 3 types of routes:
+     * - absolute with different origin: https://www.odoo.com
+     * - absolute with same origin: https://restaurant.odoo.com/menu/products
+     * - relative: /menu/products;
+     * If the route is absolute with a different origin, we redirect to it.
+     * If the route is on the same origin, we use the history API to navigate to it.
+     * (this means that we don't make aditional requests to the server)
+     */
+    navigate(route) {
+        if (!route.startsWith("/")) {
+            browser.location.href = route;
+            return;
+        }
+        const url = new URL(browser.location.href);
+        url.pathname = `menu/${this.props.pos_config_id}` + route;
```

Comment:
Prefer template strings to string concatenation. Also, do you still need to contruct a URL object? Are you still using any query paramters? A simple string should work fine here.